deluge/ui/gtkui/gtkui.py

```diff
@@ -238,36 +239,52 @@ def nsapp_open_file(osxapp, filename):
         self.rpc_stats = LoopingCall(self.print_rpc_stats)
 
         # Twisted catches signals to terminate, so have it call a pre_shutdown method.
-        reactor.addSystemEventTrigger("before", "shutdown", self.pre_shutdown)
+        reactor.addSystemEventTrigger("before", "gtkui_close", self.close)
+
         reactor.callWhenRunning(self._on_reactor_start)
+        self.closing = False
+
+        def gtkui_sigint_handler(num, frame):
+            log.debug("SIGINT signal caught - firing event: 'gtkui_close'")
+            reactor.callLater(0, reactor.fireSystemEvent, 'gtkui_close')
+
+        signal.signal(signal.SIGINT, gtkui_sigint_handler)
 
         # Initialize gdk threading
         gtk.gdk.threads_enter()
         reactor.run()
-        self.shutdown()
-        gtk.gdk.threads_leave()
-
-    def shutdown(self, *args, **kwargs):
-        log.debug("gtkui shutting down..")
-
-        component.stop()
 
         # Process any pending gtk events since the mainloop has been quit
         while gtk.events_pending():
-            gtk.main_iteration(0)
+            gtk.main_iteration_do(block=False)
+            time.sleep(0.0001)
 
+        gtk.gdk.threads_leave()
+
+    def shutdown(self, *args, **kwargs):
+        log.debug("GTKUI shutting down...")
         # Shutdown all components
-        component.shutdown()
+        if self.started_in_classic:
+            return component.shutdown()
 
+    @defer.inlineCallbacks
+    def close(self):
+        if self.closing:
+            return
+        self.closing = True
         # Make sure the config is saved.
         self.config.save()
-
-    def pre_shutdown(self, *args, **kwargs):
-        """Modal dialogs can prevent the application exiting so destroy mainwindow"""
         # Ensure columns state is saved
         self.torrentview.save_state()
+        # Shut down components
+        yield self.shutdown()
+
+        # Modal dialogs can prevent the application exiting so destroy mainwindow
+        # Must do this here to avoid hang when closing with SIGINT (CTRL-C)
         self.mainwindow.window.destroy()
```

Comment:
It is done before reactor shutdown. Before the last rebase, pre_shutdown was actually run after the close function, but now the same code is run before.